packages/arb-upgrades/index.ts

```diff
@@ -0,0 +1,361 @@
+import { HardhatRuntimeEnvironment } from 'hardhat/types/runtime'
+import { writeFileSync, readFileSync } from 'fs'
+import childProcess from 'child_process'
+import {
+  QueuedUpdates,
+  CurrentDeployments,
+  ContractNames,
+  CurrentDeployment,
+  QueuedUpdate,
+  isBeacon,
+} from './types'
+
+// import {
+//   getStorageLayout,
+//   assertUpgradeSafe,
+//   getContractVersion,
+//   assertStorageUpgradeSafe,
+//   solcInputOutputDecoder,
+//   validate,
+//   RunValidation,
+// } from '@openzeppelin/upgrades-core'
+
+const adminSlot =
+  '0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103'
+const implementationSlot =
+  '0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc'
+const beaconSlot =
+  '0xa3f0ad74e5423aebfd80d3ef4346578335a9a72aeaee59ff6cb3582b35133d50'
+
+const currentCommit = childProcess
+  .execSync('git rev-parse HEAD')
+  .toString()
+  .trim()
+
+const ensureCleanGitTree = () => {
+  const tree = childProcess.execSync('git diff-index HEAD --').toString().trim()
+  const found = tree.split('\n').find(diff => diff.includes('contracts'))
+  if (found) {
+    throw new Error(
+      `You have local changes to ${found}. Commit/stash/ get rid of them`
+    )
+  }
+}
+
+export const initUpgrades = (
+  hre: HardhatRuntimeEnvironment,
+  rootDir: string
+) => {
+  const getQueuedUpdates = async (): Promise<{
+    path: string
+    data: QueuedUpdates
+  }> => {
+    const network = await hre.ethers.provider.getNetwork()
+    const path = `${rootDir}/deployments/${network.chainId}_queued-updates.json`
+    try {
+      const jsonBuff = readFileSync(path)
+      return { path, data: JSON.parse(jsonBuff.toString()) as QueuedUpdates }
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        console.log('New network; creating queued updates file')
+        writeFileSync(path, JSON.stringify({}))
+        return { path, data: {} }
+      } else {
+        throw err
+      }
+    }
+  }
+
+  const getDeployments = async (): Promise<{
+    path: string
+    data: CurrentDeployments
+  }> => {
+    const network = await hre.ethers.provider.getNetwork()
+    const path = `${rootDir}/deployments/${network.chainId}_current_deployment.json`
+    try {
+      const jsonBuff = readFileSync(path)
+      return {
+        path,
+        data: JSON.parse(jsonBuff.toString()) as CurrentDeployments,
```

Comment:
as for the QueuedUpdates comment above, does this fail in a clean way if the json is manually altered into something incompatible?

Reply:
I'm afraid of a failure mode similar to something described [here](https://basarat.gitbook.io/typescript/type-system/type-assertion#:~:text=but%20the%20hazard%20here%20is%20the%20same%2C%20if%20you%20forget%20a%20property%20the%20compiler%20will%20not%20complain.%20It%20is%20better%20if%20you%20do%20the%20following%3A)